File: Palind/palind.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int n, p, q;
char same[13][2];
char input[51] = {};

void processInput();

int main(void)
{
	scanf("%d\n", &n);
	for (int loop = 1; loop <= n; loop++)
	{
		scanf("%d\n", &p);
		int i;
		for (i=0; i < p; i++)
			scanf("%c %c\n", &same[i][0], &same[i][1]);

      printf("Test case #%d:\n", loop);

		scanf("%d\n", &q);
		for (i=0; i < q; i++)
			processInput();
         
      printf("\n");
	}

	return 0;
}

void processInput()
{
	char tupin[51] = {};
	int i, j;
	scanf("%s", input);

	printf("%s ", input);
	for (i = 0; i < strlen(input); i++)
		for (j = 0; j < p; j++)
			if (input[i] == same[j][0])
				input[i] = same[j][1];

	for (i = 0; i < strlen(input); i++)
		tupin[i] = input[strlen(input)-1-i];

	if (strcmp(tupin, input) == 0)
		printf("YES\n");
	else 	printf("NO\n");
}

File: Palind/Palind.java
import java.io.*;
import java.util.*;

public class Palind
{

	HashMap<Character, Character> same;
	
	public Palind(String filename)
	{
		try (Scanner in = new Scanner(new File(filename)))
		{
			int testCases = in.nextInt();
			
			for (int a = 1; a <= testCases; a++)
			{
				// read in pairs of phonemes
				int numPairs = in.nextInt();
				same = new HashMap<Character, Character>();
				
				for (int i = 0; i < numPairs; i++)
				{
					// store (c1, c2) and (c2, c1) so we can loop up both ways
					char c1 = in.next().charAt(0);
					char c2 = in.next().charAt(0);
					
					same.put(c1, c2);
					same.put(c2, c1);
				}
					
				
				System.out.println("Test case #" + a + ":");
				
				// read in q strings to test
            int q = in.nextInt();
            in.nextLine();
				for (int i = 0; i < q; i++)
				{
					String str = in.nextLine().trim();
					System.out.println(str + " " + (isPhonemePalindrome(str) ? "YES" : "NO"));
				}
					
				
				System.out.println();
			}
			
		}
		catch (FileNotFoundException e)
		{
			e.printStackTrace();
		}
	}
	
	public boolean isPhonemePalindrome(String str)
	{
		for (int i = 0, j = str.length()-1; i < j; i++, j--)
		{
			char c1 = str.charAt(i);
			char c2 = str.charAt(j);
			
			if (c1 != c2)
				if (same.get(c1) == null || same.get(c1) != c2)
					return false;
		}
		
		return true;
	}
	
	public static void main(String[] args)
	{
		new Palind("palind.in");
	}
}
